Add new filter 'mu_plugins'

nate-allen opened the pull request:
Adds a new filter called `mu_plugins` that allows the filtering of the must-use plugins array.

Trac ticket: https://core.trac.wordpress.org/ticket/57278

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

nate-allen committed:
Add new filter 'mu_plugins'

mukeshpanchal27 commented on src/wp-admin/includes/plugin.php:
```diff
@@ -407,7 +407,14 @@ function get_mu_plugins() {
+	 *
+	 * @since 6.2.0
+	 *
+	 * @param array $wp_plugins An array of arrays of mu-plugin data, keyed by plugin file name.
```
```suggestion
	 * @param array[] $wp_plugins An array of arrays of mu-plugin data, keyed by plugin file name.
```

mukeshpanchal27 reviewed (commented):
Thanks @nate-allen, Left one nit-pick document changes. Can you please add unit tests for the new filter?

nate-allen committed:
Add brackets after 'array' in docblock

nate-allen committed:
Merge branch 'trunk' into trac-57278

nate-allen committed:
Remove 'mu_plugins' filter

After some discussion on the trac ticket, it was suggested that
a general purpose filter would be of more use.

See: https://core.trac.wordpress.org/ticket/57278#comment:6

nate-allen committed:
Add new filter 'plugins_list'

nate-allen committed:
Add test for plugins_list filter

nate-allen committed:
Merge branch 'trunk' into trac-57278

nate-allen committed:
Prevent TypeError by checking for null before strlen

This commit updates the code to check for null before calling strlen()
to prevent a TypeError in PHP 8.

costdev commented on tests/phpunit/tests/admin/wpPluginsListTable.php:
```diff
@@ -56,4 +75,31 @@ public function test_get_views_should_return_views_by_default() {
 	}
+
+	/**
+	 * @ticket 57278
```
```suggestion
	 * Tests that WP_Plugins_List_Table::prepare_items() 
	 * applies 'plugins_list' filters.
	 *
	 * @ticket 57278
	 *
	 * @covers WP_Plugins_List_Table::prepare_items
```

costdev commented on tests/phpunit/tests/admin/wpPluginsListTable.php:
```diff
@@ -56,4 +75,31 @@ public function test_get_views_should_return_views_by_default() {
+		$this->assertSame( $plugins, $this->fake_plugin );
+	}
+
+	public function plugins_list_filter( $plugins_list ) {
```
```suggestion
	/**
	 * Adds a fake plugin to an array of plugins.
	 *
	 * Used as a callback for the 'plugins_list' hook.
	 *
	 * @return array
	 */
	public function plugins_list_filter( $plugins_list ) {
```

costdev commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -292,11 +292,20 @@ public function prepare_items() {
 		}
 
-		if ( strlen( $s ) ) {
+		if ( isset( $s ) && strlen( $s ) ) {
```
This change doesn't appear to be related to the addition of the new filter. @nate-allen Was this part of a patch for another ticket?

nate-allen replied:
Hi @costdev. It's required for the unit test for this ticket.

costdev replied:
Thanks @nate-allen!

Pinging @TimothyBJacobs and @spacedmonkey as REST API maintainers. Are you happy for this source change to be made here to allow PHPUnit tests to pass, or should it be made in a separate ticket, or should the test(s) instead set a value for the `$s` global?

TimothyBJacobs replied:
@costdev This doesn't look REST API related to me? But since I'm here 🙂.

To me it looks like the relevant globals should be getting initialized in the test. Though I think @SergeyBiryukov or @hellofromtonya could give a more informed opinion on the direction for these kinds of unit tests.

costdev replied:
Thanks @TimothyBJacobs! It appears my brain skipped a beat when I posted that at 3AM 😅

Yeah that's what I was thinking - and that the `isset( $s )` check, if it's going to be added at some point, should probably be separate rather than bundled in the addition of a new filter.

Let's see what Sergey or Tonya think too 🙂

hellofromtonya replied:
Thanks for ping @TimothyBJacobs and @costdev.

>To me it looks like the relevant globals should be getting initialized in the test.

In this PR, yes, I agree. 

How to do it?

- Store each global's original value in a new property added to the test class. 
    - Add the property.
    - The initialization of the property in either the `before_set_up_class()` or `set_up()` fixture method.
- The test is then able to modify the global to its needs.
-  How to restore / reset the original value? In the `tear_down()`, set the global to its the original value using the property. 
    - Why the `tear_down()` and not within the test method? The `tear_down()` ensures the global gets reset, as it runs after each test regardless if the test fails and skips any of it test code.

[Here's an example test](https://github.com/WordPress/wordpress-develop/blob/trunk/tests/phpunit/tests/webfonts/wpThemeJsonWebfontsHandler.php).

nate-allen replied:
Thank you @hellofromtonya (and everyone else that has helped move this along!) I will make these changes and reach out if I have any issues.

costdev reviewed (changes requested):
Thanks for the PR @nate-allen! I've left some feedback in this review.

nate-allen committed:
Update docblock

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

nate-allen committed:
Update docblock

Co-authored-by: Colin Stewart <79332690+costdev@users.noreply.github.com>

mukeshpanchal27 commented on src/wp-admin/includes/class-wp-plugins-list-table.php:
```diff
@@ -292,11 +292,20 @@ public function prepare_items() {
+		/**
+		 * Filters the array of plugins for the list table.
+		 *
+		 * @since 6.2.0
```
```suggestion
		 * @since 6.3.0
```

Nit-pick

nate-allen committed:
Update @since version

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

audrasjb committed:
Merge branch 'trunk' into trac-57278

audrasjb commented:
I committed fb4a416 to solve a conflict against trunk.

audrasjb committed:
Fix WPCS issues

audrasjb committed:
Another WPCS fix

audrasjb committed:
Final WPCS fix ^^

audrasjb commented:
closing in favor of #4717
